src/chains/ethereum/ethereum/src/block-time.ts

```diff
@@ -0,0 +1,100 @@
+export interface BlockTime {
+  /**
+   * Set the current time of the BlockTime instance. Mining immediately afterwards will result in a block timestamp of ~timestamp (some milliseconds could pass before the block is mined).
+   * @param  {number | Date} timestamp - The time that should be considered "current time".
+   * @returns number - The "offset" for the BlockTime instance in milliseconds.
+   */
+  setTime(timestamp: number | Date): number;
+
+  /**
+   * Get the current offset of the BlockTime instance. This is the value that would be added to the system time in order to result in the desired timestamp.
+   * @returns number - The "offset" for the BlockTime instance in milliseconds.
+   */
+  getOffset(): number;
+
+  /**
+   * Set the current offset of the BlockTime instance. This is the value that would be added to the system time in order to result in the desired timestamp.
+   * @param {number} offset - The "offset" for the BlockTime instance in milliseconds.
+   */
+  setOffset(offset: number): void;
+
+  /**
+   * Create a timestamp for a new block in seconds. Can specify the timestamp for the block, in milliseconds, which will be reflected in future calls to this function.
+   * @param  {number} timestamp? - Optional - The timestamp for the block that will be mined, in milliseconds.
+   * @returns number - the timestamp in seconds
+   */
+  createBlockTimestampInSeconds(timestamp?: number | Date): number;
+}
+
+export class ClockBasedBlockTime implements BlockTime {
+  private _getReferenceClockTime: () => number | Date;
```

Comment:
I changed all of the # prefixed private class fields to _ prefixed typescript privates. The former have performance implications, and I don't see the benefit until we can resolve this.

#3240 addresses this more generally.